ZEP Build State on Followers (#16)
zeebe-io/enhancements: ZEP000-build-state-on-followers.md

[comment]: <> (           * To compact our log, we need to be able to take snapshots from the state.)
[comment]: <> (             * The state need to contain a dense reflection of the log, last exported and processed position, from which we can recover.)
[comment]: <> (           * To take a snasphot from the state, we need to build the state on all participants &#40;leader and follower&#41; in order to not replicate it.)
[comment]: <> (           * To build state on followers, we need to avoid creating follow-up events by the state machine, Single Writer Principle. )
[comment]: <> (           * To avoid follow-up events, we need to either replay only or use a NoopWriter.)
[comment]: <> (In the following sections we will go in more depth in some these mentioned parts of the chain.)

### Compacting the log

In order to avoid an ever-growing log we need to compact it from time to time. This can be done by taking snapshots from the state, which allows us to start from that specific snapshot, such that we can compact our log. With building state on followers each node is responsible for taking their own snapshots and compacting their log.

> Part of the snapshot should be the last processed and last exported position for each exporter. The smallest number indicates until which position we can compact our log. In order to avoid running the exporters on all nodes and reduce the network load (since they export normally to an external system), the Leader has to periodically sync the last exported position to the followers. This is done via SWIM.

Reviewer: What's wrong with producing an event on the log which can be applied to update the position? I see it mentioned in alternatives, but there's no explanation as to why it was discarded?

> Part of the snapshot should be the last processed and last exported position for each exporter. The smallest number indicates until which position we can compact our log. In order to avoid running the exporters on all nodes and reduce the network load (since they export normally to an external system), the Leader has to periodically sync the last exported position to the followers. This is done via SWIM.

Reviewer 2: I agree that we should discuss this point a bit more (in the ZEP).

> Part of the snapshot should be the last processed and last exported position for each exporter. The smallest number indicates until which position we can compact our log. In order to avoid running the exporters on all nodes and reduce the network load (since they export normally to an external system), the Leader has to periodically sync the last exported position to the followers. This is done via SWIM.

Author: > Does it have to be SWIM, or should we send it via direct leader-follower communication via unicast or broadcast?

No doesn't need to be SWIM. I was also not sure to mention this here. It is one way to do it.

> What's wrong with producing an event on the log which can be applied to update the position? I see it mentioned in alternatives, but there's no explanation as to why it was discarded.

For me the sending of the state sound it simple enough for the first solution. It doesn't mean we can't improve. I can imagine that we could also iterate over it later and go with another solution. If you go directly with it it means we need to introduce a new record or value type (lot of protocol changes), have a writer in the exporter, and also a replay mechanism for the exporter, we need to ignore exporter records etc.

> Part of the snapshot should be the last processed and last exported position for each exporter. The smallest number indicates until which position we can compact our log. In order to avoid running the exporters on all nodes and reduce the network load (since they export normally to an external system), the Leader has to periodically sync the last exported position to the followers. This is done via SWIM.

Author: > I agree that we should discuss this point a bit more (in the ZEP).

I would like to start first with the prototype and need to find out how it could really work. Maybe I see then some issues with the proposed solution, which I currently not seeing.

But in general I think this ZEP should focus on building state, which is highly focuses on the processing part and could be seen as the base. After that we can still improve and write a new ZEP which covers the exporter part more in depth. It is just a matter of being focused I would say. The exporter stuff has a rat tail. Otherwise I fear we spent to much time to figure it out all it once.

> The last processed position corresponds to the last processed command on the leader **or** the last applied event on the follower. 

Author: Ah good point. I think it should be source position of the event right?

  * Bootstrapping Leader Partition, no already running partition
  * Bootstrapping Follower Partition, no already running partition
  * Switch to Leader Partition, from an already running Follower partition
  * Switch to Follower Partition, from an already running Leader Partition.

In the following section we will use the term processing and replay heavily, please read the [ZEP004](https://github.com/zeebe-io/enhancements/blob/master/ZEP004-wf-stream-processing.md) if you want to know more about it.

### Bootstrap Leader Partition

![stateOnFollower-BootstrapLeader](images/stateOnFollower-BootstrapLeader.png)

In this scenario we have no already running Zeebe partition and installed it completely new. During installing our services we restore the state, via copying the latest snapshot into the runtime folder. After installing all services, we start with the "Stream Processing", which is on the Leader separated in two parts. The first part is the processing, which consist of the following steps:

  * **Replay until end**, here we replay all existing events on the Log, which were not part of the Snapshot. This is necessary to get the latest state, before processing.

>   * **Reset Dispatcher Position**, here we set the Dispatcher position which generates the position for our follow-up events.

Reviewer: I guess we generally want to reset the dispatcher entirely, no? I'd be a bit worried if it's long living of having left over stuff in there.

Author: Yes we should wipe it right.

[comment]: <> (       - [ ] Does the ZEP affect the official Zeebe distribution &#40;e.g. configuration, logging&#41;?)
[comment]: <> (       - [ ] Does the ZEP require coordination with the platform team?)
[comment]: <> (       - [ ] Does the ZEP require coordination with the Operate team?)

**TODO:** After doing prototype rewrite this section.

We can't reuse our current ReprocessingStateMachine, since it is written in a way such it doesn't support continuously replay. Currently, it is a one time thing. It scans the log for the last source position to determine, when it will stop with the replay. Furthermore, it collects error events to make sure the records are failing on reprocessing as well. Should be fixed with https://github.com/camunda-cloud/zeebe/issues/7270. We need a new ReplayStateMachine, which only consumes events continuously. The leader is the only one who is allowed to write commands and follow-up events, which ensures that the follower will not be faster than the leader.

       
## Compatibility

> We see no issues with this proposal.

Reviewer 2: How to handle different versions between leader and follower? 
At some point, we want to support rolling updates. We should have an idea on how to handle these situations.

Author: Good Point. I probably need to add a section for that. But I'm also not sure whether we know how we handle it currently. :sweat_smile:

Reviewer 3: Isn't that also a problem currently?

Author: @deepthidevaki I thought as well but was not sure :sweat_smile:

Reviewer 2: Yes. It is already a kind of problem if there is an unlucky leader change during the update. 
But with this change, we will run definitely into this problem because all nodes process/replay the records.

        
### Unit

### Integration

  * We need to verify that the state is build on Leaders **AND** follower's. 
  * We need to make sure that the followers take snapshots and can compact their log.
  * We need to verify that the blacklisting is build up correctly on the follower side and that if it becomes leader it ignores related commands.
  * In general, we should verify that fail-over works as expected and that the followers are building there state correctly, such that a new Leader can use it.
  * After fail-over the new Leader should be able to export after the last lowest exporter position.
  * After fail-over the new Leader should be able to process after the last processed command.
  * InstallRequest's should be properly handled by followers, and they should continue with applying events after restoring the state.

### E2E

 In the E2E test we should verify similar things:

 * We should verify that fail-over work as expected and that the followers are building there state correctly. The leaders should be able to continue where the followers left of and there should be no issue on exporting records from the new leader.

### Benchmarks

> Verify the improvements in process execution latency during fail-over and also the latency from Follower-To-Leader processing transition, which was the main goal.

Reviewer 2: Do you have any new metrics in mind that could be useful?

Author: Most of it we have already. Like when it starts to process again. But I had a new idea for a metric. The time when the leader role event arrives at our Zeebe Partition until the time the Stream processor starts to process. This is actually the time we want to minimized in that ZEP.

# Drawbacks
[drawbacks]: #drawbacks

> Time? 

Reviewer 2: :sweat_smile: I think one drawback of this approach is that the followers need to do more stuff. Previously, a follower just receives a new snapshot. Now, a follower replays all events to build the state. I assume that as a consequence a follower uses more resources.

[comment]: <> (      This section is intended to encourage you as an author to think about the lessons from other languages, provide readers of your ZEP with a fuller picture. If there is no prior art, that is fine - your ideas are interesting to us whether they are brand new or if it is an adaptation from other languages.)
[comment]: <> (      Note that while precedent set by other products is some motivation, it does not on its own motivate a ZEP.)

There are several raft implementations out there and all of them doing it in the same way, the state machine is running on all nodes and build the state and take snapshot and compact the log independently.

Other implementations:

  * Hashicorp raft https://github.com/hashicorp/raft
  * etcd raft https://github.com/etcd-io/etcd/tree/master/raft
  * dragonboat https://github.com/lni/dragonboat

We also saw [above](#distinction-to-normal-raft) the Raft paper, which mentions that this is the way how we should implement it.

# Out of scope
[out-of-scope]: #out-of-scope

[comment]: <> (Call out anything which is explicitly not part of this ZEP.)

Building and supporting large states was not covered by this ZEP.

> # Unresolved questions

Reviewer: What will the performance characteristics be?

Author: Should I add this as question?

---
title: Build state on followers
authors:
- Christopher Zell (@zelldon)
  reviewers: []
  approvers: []
  editor: TBD
  creation-date: 2020-04-28
  last-updated: 2021-06-14
  status: provisional
  see-also: []
  replaces: []
  superseded-by: []
---

# Summary
[summary]: #summary

We should build state on followers to be more reliable, future proven and enable fast fail-over.

# Motivation
[motivation]: #motivation

> Generally we want to reduce the failover time. With that we mean: the time between no leader is available, processing has stopped, and a new leader has been chosen and processing has started again. Especially we want to minimize the impact on process execution latency during failover on a partition. With this ZEP we want to decrease the time from node a becomes Leader until it starts to process new commands. This can be achieved with building state on followers.

Reviewer 4: 🙃 
```suggestion
Generally we want to reduce the failover time. With that we mean: the time between no leader is available, processing has stopped, and a new leader has been chosen and processing has started again. Especially we want to minimize the impact on process execution latency during failover on a partition. With this ZEP we want to decrease the time from when a node becomes Leader until it starts to process new commands. This can be achieved with building state on followers.
```

departure is justified. While having a leader helps avoid
conflicting decisions in reaching consensus, consensus
has already been reached when snapshotting, so no decisions conflict. Data still only flows from leaders to fol
lowers, just followers can now reorganize their data.
>
> We considered an alternative leader-based approach in
which only the leader would create a snapshot, then it
would send this snapshot to each of its followers. However, this has two disadvantages. First, sending the snapshot to each follower would waste network bandwidth and
slow the snapshotting process. Each follower already has
the information needed to produce its own snapshots, and
it is typically much cheaper for a server to produce a snapshot from its local state than it is to send and receive one
over the network. Second, the leader’s implementation
would be more complex. For example, the leader would
need to send snapshots to followers in parallel with replicating new log entries to them, so as not to block new
client requests.

As the paper states it makes totally sense to build state and take snapshots of it on all servers unrelated of the raft state.

> Unfortunately our processing state machine produces new commands, which need to be applied to the state machine again. This means that we can't just run the same processing state machine on all nodes, since only the leader should be able to create new records.

Author: In normal RAFT system you have one statemachine which you can run on all machines, no matter what type/role the node has. In our case this is not true, since only the leader should create follow up events. So we have a distinction between processing and replay state machine. The first one runs on the leader the other on the follower.

> Unfortunately our processing state machine produces new commands, which need to be applied to the state machine again. This means that we can't just run the same processing state machine on all nodes, since only the leader should be able to create new records.

Reviewer 4: But how does that stop us from **making snapshots on followers**?

> Unfortunately our processing state machine produces new commands, which need to be applied to the state machine again. This means that we can't just run the same processing state machine on all nodes, since only the leader should be able to create new records.

Author: Why does it stop us? The text just says that we can;t just run the same state machine, the same logic to build it.

> Unfortunately our processing state machine produces new commands, which need to be applied to the state machine again. This means that we can't just run the same processing state machine on all nodes, since only the leader should be able to create new records.

Reviewer 4: Ah I see, the problem is that you're describing 2 different topics in this subchapter.

The heading of this subchapter is `Distinction to normal raft`. If we can make snapshots on followers, then that's actually the same as in the paper right? Perhaps it just doesn't belong here, but in a separate heading called `Snapshots on followers`.

> Unfortunately our processing state machine produces new commands, which need to be applied to the state machine again. This means that we can't just run the same processing state machine on all nodes, since only the leader should be able to create new records.

Author: > If we can make snapshots on followers, then that's actually the same as in the paper right? 

It is not the same, since we can't just run the same code which we run on the Leader :)

> Unfortunately our processing state machine produces new commands, which need to be applied to the state machine again. This means that we can't just run the same processing state machine on all nodes, since only the leader should be able to create new records.

Author: I can also remove it  when it lead to more confusion than it helps. Any opinion @deepthidevaki

> Unfortunately our processing state machine produces new commands, which need to be applied to the state machine again. This means that we can't just run the same processing state machine on all nodes, since only the leader should be able to create new records.

Reviewer 4: Well it makes sense to discuss the topic of snapshots on followers to some extend. Is it in scope of this ZEP or out of scope? If it's out of scope you could also mention it in that section,

> Unfortunately our processing state machine produces new commands, which need to be applied to the state machine again. This means that we can't just run the same processing state machine on all nodes, since only the leader should be able to create new records.

Reviewer 3: I tried to rewrite it to make it more clear. Does that help?

```
As the paper states it makes totally sense to build state and take snapshots of it on all servers unrelated of the raft state.

However, we cannot run the same state machine on the leader and followers as in the original raft.
Our processing state machine produces new follow up commands.The order in which follow up commands and events are written to the logstream in relation to concurrent external commands is decided by the leader. When the follower runs the same state machine, it will also generate new follow up commands and events, which might result in a different order of records. This means that we can't just run the same processing state machine on all nodes, since only the leader should be able to create new records. As later explained in this document, the leader's state machine process commands and generate followup commands and events. Follower runs a replay state machine where it only applies the events generated by the leader.
```

> Unfortunately our processing state machine produces new commands, which need to be applied to the state machine again. This means that we can't just run the same processing state machine on all nodes, since only the leader should be able to create new records.

Reviewer 4: I still don't see what this has to do with snapshots on followers. It seems to describe the difference between a leader (processes: writes follow-up records) and a follower (only replays), but it does not discuss the impact on snapshots on those followers. 

At this point it is still unclear to me whether snapshots on followers are in scope of the ZEP or not

> Unfortunately our processing state machine produces new commands, which need to be applied to the state machine again. This means that we can't just run the same processing state machine on all nodes, since only the leader should be able to create new records.

Author: I have the feeling you are missing some understanding of the ZEP. In general the ZEP is about building state on Followers, in order to do that we replay events on the followers. Taking snapshot on Followers is just a part of it.

> Unfortunately our processing state machine produces new commands, which need to be applied to the state machine again. This means that we can't just run the same processing state machine on all nodes, since only the leader should be able to create new records.

Author: I'm wondering why this is the case. What I have missed to make this clear? Or what do you think we should add?

> Unfortunately our processing state machine produces new commands, which need to be applied to the state machine again. This means that we can't just run the same processing state machine on all nodes, since only the leader should be able to create new records.

Reviewer 4: I understand that the ZEP is about state on followers. and so it also makes sense to discuss snapshots because of the reduced complexity and network traffic related to distributing snapshots if you only make them on leaders.

To me the part up to and including the sentence `As the paper states it makes totally sense to build state and take snapshots of it on all servers unrelated of the raft state.` was clear. 

The paragraph under it, then continues with `Unfortunately` implying that even though it seems possible to create snapshots on followers, it is not. And then continues to talk about the difference between leaders and followers concerning the difference of processing and replay without actually stating how this affects snapshots. This paragraph just confuses me about what was mentioned before it. And I think the reason is that they just discuss different topics (as far as I understand now, the last paragraph has nothing to do with snapshots)

> Unfortunately our processing state machine produces new commands, which need to be applied to the state machine again. This means that we can't just run the same processing state machine on all nodes, since only the leader should be able to create new records.

Reviewer 4: Sorry to be so difficult 😄 I just want to understand it

> Unfortunately our processing state machine produces new commands, which need to be applied to the state machine again. This means that we can't just run the same processing state machine on all nodes, since only the leader should be able to create new records.

Reviewer 3: To take snapshot on followers, we need to build state on followers. To build state on followers, we need to run a state machine on the followers. If we could follow the normal raft approach, we can run the same state machine on the leader and follower. Then we don't need this ZEP, as it is trivial to build state and take snapshots. So the main difference compared to raft is how to build state on followers. Once we know how to build state on followers, taking snapshot is not difficult.

> Unfortunately our processing state machine produces new commands, which need to be applied to the state machine again. This means that we can't just run the same processing state machine on all nodes, since only the leader should be able to create new records.

Author: Thanks @deepthidevaki that is exactly it

> Unfortunately our processing state machine produces new commands, which need to be applied to the state machine again. This means that we can't just run the same processing state machine on all nodes, since only the leader should be able to create new records.

Reviewer 4: Thanks @deepthidevaki that makes sense. Can we add this to this subchapter?

> Unfortunately our processing state machine produces new commands, which need to be applied to the state machine again. This means that we can't just run the same processing state machine on all nodes, since only the leader should be able to create new records.

Author: If you think it helps sure. I will change and apply all the hints after all the reviews are done, such that the other reviews haven't any conflicts on reading.

> Unfortunately our processing state machine produces new commands, which need to be applied to the state machine again. This means that we can't just run the same processing state machine on all nodes, since only the leader should be able to create new records.

Author: https://github.com/zeebe-io/enhancements/pull/16/commits/c521577fa1cea096df23157ea95af55c5d43b10d

## State on Followers

Instead of running the same processing state machine on the followers, we replay the events from the replicated log. For that we have a ReplayStateMachine, which allows to continuously replay events and apply them to the state. We call it replay, since they have been already applied on the leader side and have been produced by him. Furthermore, we will reuse this state machine on bootstrap a leader partition such that the term "replay" makes sense here as well.

Raft ROLEs are transient states, which means it is likely that a Leader change happens. How the system reacts on these role changes can be seen in the following picture. With the following process, we minimize the time for the Follower-to-Leader transition, which impacts the process execution latency.

> ![stateOnFollower](images/stateOnFollower.png)

Author: https://github.com/zeebe-io/enhancements/pull/16/commits/7527f8181b61edc154b995f00635b62b5cedb56b I reworked it

Part of the Follower "Stream Processing" is the continuously applying of events, which is called **Replay Events**, since they already applied on the Leader. This avoids generating follow-up events on the follower. Only the Leader is allowed to write records, Single Writer Principle (SWP). Additionally, we can make sure that the follower is not faster than the leader, and we can easily switch to the Leader processing as we can see in the next section.

In distinction to the Leader, the Follower will not execute any exporters. It will receive periodically the last successful exported positions, which he stores in his state.

It can happen, if the follower is slow, or was not available for longer time that it receives an "InstallRequest" from the Leader. Such an "InstallRequest" contains the latest snapshot from the Leader. This is done to reduce the time the follower needs to catch up, or if the leader already has compacted his log. This scenario is described later [here](#re-init-follower-partition).

> One of an edge case topic is the "Blacklisting" of process instances. This done on the Leader side, when during processing an error occurs. The related process instance, where the command corresponds to and caused this error, will be "Blacklisted". To persist that, an error event is written to the log. This kind of error events are applied on the follower replay mode to restore the blacklisted process instances. This is necessary such that we ignore on normal processing/applying of events related commands/events, otherwise we might end in an unexpected state. This restoring of the "Blacklist", is also done on the Leader replay mode.

Author: Edge case in a sense it is handled differently before and happens (or at least should) rarely

> One of an edge case topic is the "Blacklisting" of process instances. This done on the Leader side, when during processing an error occurs. The related process instance, where the command corresponds to and caused this error, will be "Blacklisted". To persist that, an error event is written to the log. This kind of error events are applied on the follower replay mode to restore the blacklisted process instances. This is necessary such that we ignore on normal processing/applying of events related commands/events, otherwise we might end in an unexpected state. This restoring of the "Blacklist", is also done on the Leader replay mode.

Reviewer 4: >handled differently before

Are error events replayed differently? Can you explain what you mean?

> One of an edge case topic is the "Blacklisting" of process instances. This done on the Leader side, when during processing an error occurs. The related process instance, where the command corresponds to and caused this error, will be "Blacklisted". To persist that, an error event is written to the log. This kind of error events are applied on the follower replay mode to restore the blacklisted process instances. This is necessary such that we ignore on normal processing/applying of events related commands/events, otherwise we might end in an unexpected state. This restoring of the "Blacklist", is also done on the Leader replay mode.

Author: @korthout no before we had different logic. We scan the complete log, collect all error events and related positions of the commands. We process the command again to make sure that it failed again. Later we just blacklisted the PI with the command. With this ZEP we want to handle them like normal events and replay them.

See related PR's:

 * https://github.com/camunda-cloud/zeebe/pull/7440
 * https://github.com/camunda-cloud/zeebe/pull/7446

  
[comment]: <> (          We have several challenges we need to overcome to solve/implement build state on followers. Let's take a look of the following causality chain:)
[comment]: <> (           * To have a working and long living system, we need to be able to compact our log no matter the role. )
[comment]: <> (           * To compact our log, we need to be able to take snapshots from the state.)
[comment]: <> (             * The state need to contain a dense reflection of the log, last exported and processed position, from which we can recover.)
[comment]: <> (           * To take a snasphot from the state, we need to build the state on all participants &#40;leader and follower&#41; in order to not replicate it.)
[comment]: <> (           * To build state on followers, we need to avoid creating follow-up events by the state machine, Single Writer Principle. )
[comment]: <> (           * To avoid follow-up events, we need to either replay only or use a NoopWriter.)
[comment]: <> (In the following sections we will go in more depth in some these mentioned parts of the chain.)

### Compacting the log

> In order to avoid an ever-growing log we need to compact it from time to time. This can be done by taking snapshots from the state, which allows us to start from that specific point, such that we can compact our log. With building state on followers each node is responsible for taking their own snapshots and compacting their log.

Reviewer 4: 🔧  In the guidelevel explanation it makes sense to quickly mention what log compaction actually is. In addition, let's add what we can achieve with this.
```suggestion
In order to avoid an ever-growing log we need to compact it from time to time. This can be done by taking snapshots from the state, which allows us to start from that specific point and compact the log (i.e. delete exported records from the log preceding the snapshot). By building state on followers each node can be responsible for taking their own snapshots and compacting their log. This reduces network traffic, because snapshots no longer need to be replicated.
```

> Part of the snapshot should be the last processed and last exported position for each exporter. The smallest number indicates until which position we can compact our log. In order to avoid running the exporters on all nodes and reduce the network load (since they export normally to an external system), the Leader has to periodically sync the last exported position to the followers. This is done via netty.

Reviewer 4: 🔧 Let's add why the exported position for each exporter matters

```suggestion
Part of the snapshot should be the last processed and last exported position for each exporter. Records must be exported by all exporters before we can delete them, so the smallest number indicates the position until which we can compact our log. In order to avoid running the exporters on all nodes and reduce the network load (since they export normally to an external system), the Leader has to periodically sync the last exported position to the followers. This is done via netty.
```

> The last processed position corresponds to the last processed command on the leader **or** to the source position of last applied event on the follower.

Reviewer 4: ❌ This is only true for the follower that is helping the quorum. This is not true for slow followers.

Author: Hey @korthout this doesn't matter whether it is a slow follower or not :)

Author: Here it is just about the mapping what means last process position for leader (processing state machine) and what for follower (replay state machine) to take a snapshot.

One important part before we start with the Follower "Replay Mode" is that we need to restore the current state from the latest snapshot. This is necessary because when we process commands we immediately apply the follow-up events to our state, which we write to the log later. We can't be sure what we have already applied is committed, and to not apply events twice it is easier to just restore from the latest snapshot, KISS principle. Other than that it is similar to described [above](#bootstrap-follower-partition).

### Re-init Follower Partition

This can be seen as a special case and can also be called as Follower-To-Follower transition. In RAFT it can happen that InstallRequests are send to an slow Follower or to one which is lagging behind. In order to get the Follower back to the latest state, the Leader will send the Follower his latest snapshot.

> The Follower will reset his log, such that no gaps in the log exist. The Leader will send log entries after the snapshot index to the follower only. In order to replay with the latest state and on the correct index and position it is necessary to completely reset the follower. The easiest way is to trigger a new follower transition, which is similar to [Bootstrap the Follower partition](#bootstrap-follower-partition). We just need to make sure that the received snapshot is copied into the runtime folder.

Author: Thanks I added this as unresolved question https://github.com/zeebe-io/enhancements/pull/16/commits/4923ce72b54905af18d185a70a8bde237ad6b1e0

## Replay State Machine

In this section we will go into more detail how the so called `ReplayStateMachine` should look like.
We will describe the functionality and how it works based on the following process model. The model is a simplified version. For example, the "Read Next Event" covers some details like filtering out records which are not events etc.

![replayStateMachine](images/replayStateMachine.png)

When starting the state machine we will first seek to the snapshot position (or start from begin if there is none). After that we try to read the next event, filtering out other types of records. If there is no event, which can be applied, then we will check whether we are in a continuous replay mode or not.

> On a Leader we just want to replay until the end of the log, to replay all remaining events to build up the latest state and then end the replay. Afterwards the Leader will go into the processing mode.

Author: Yes you're right thanks. I added more details. https://github.com/zeebe-io/enhancements/pull/16/commits/826d28675822bcf8496c8db26c3ff4bfd51d907a

```
AppendRequest {
  long term
  String leader
  long prevLogIndex
  long prevLogTerm
  List<PersistedRaftRecord> entries
  long commitIndex
}
```

An [AppendRequest](https://github.com/camunda-cloud/zeebe/blob/develop/atomix/cluster/src/main/java/io/atomix/raft/protocol/AppendRequest.java) can contain multiple raft entries. If the quorum is reached for a certain index, the index can be committed. The last committed index is shared together with the raft entries via an `AppendRequest`.

![leaderCommit](images/leaderCommit.png)

If the Leader commits an index it can immediately call the RAFT commit listeners with the corresponding `IndexedRaftEntry`. This is necessary, since we need to translate the index of the committed entry into a committed position later. We can do this in the `AtomixLogStorage` via asking for the highest position, if the entry contains application entries. With this highest position we can set the commit position in our LogStream abstraction. The LogStream will then notify all our other commit listeners, which wait for the committed position.

On the follower it looks similar, but it works a bit different.

![followerCommit](images/followerCommit.png)

> As written above, the follower receives `AppendRequests`, which will contain the last committed index. The committed index, does need to correspond to the entries which are sent via the `AppendRequest`. In order to be able to translate the index to the position, we need the `IndexedRaftEntry`. For that the Follower has a separate reader, which is always forwarded to the next committed index to get the latest committed IndexedRaftEntry, which he then uses to call the RAFT `CommitListeners`. The rest stays the same as on the Leader side.

Reviewer 3: As I said before, I'm not a fan of having another reader for the sake of commit listener. But I wouldn't see it as a blocker for now. We can always improve if we find a better solution.

Author: Agree. Alternative would be https://github.com/zeebe-io/enhancements/pull/16#discussion_r661751155

Author: I think we will go with the proposed solution in the comment, which allows to avoid a separate reader. I will rewrite this section.

> 

Author: I thought longer about this part, after @npepinpe mentioned he would like to have a better split of concerns and that the `LogStream` (at least the consumers) doesn't know about commits at all.

I have an alternative idea to the commit listener stuff, if people like it, we could use it instead.

## LogStream

The interface will be changed such it supports a new listener concept, where you can register for new records becoming available, internally this means it is committed (`LogStreamImpl` still uses the commit listeners) .

```java

public interface LogStream {

  public void listenForNewRecordAvailable(Runnable listener);

  public ActorFuture<Void> awaitRecordWithPosition(long position);

}
```

The runnable/listener we register on the `LogStream` is called, when internally the next commit index is increased and can do with that what ever it wants. For example try to read the next record.

There are some cases where we want to listen or wait for a specific position, like on leader snapshotting and error record writing. Both happen only on the Leader side. The follower side hasn't such a use case. This means we could either have a separate `LogStream` implementation for the Follower, which throws unsupported exception or have internal (in the `LogStreamImpl`) a conditional switch. In order to achieve this we could reuse the append listeners (we currently use).

For the snapshotting I would propose to have two implementations, since on the Follower side we can immediately mark it as committed/valid. There is no need to wait for a position. On the Leader side we have a SnapshotDirector implementation which awaits a certain position, we no longer need to implement the commit listener here and just react on the future completion.

On the `ProcessingStateMachine` we can also await that the error record with the position we wrote it is available.

In general I think this solution is a bit more work to implement. It means we have more separations of certain classes, like having different SnapshotDirector or different LogStream (e.g. writable and read only logstream), but it might help in understanding the different Roles and approaches.

Let me know what you think. Especially @npepinpe. Might be a crazy idea and it might be just too late for me.

P.S.: I currently see no way around on not waiting for a certain position. In the snapshot director we could wait that the last processed position reaches the last written position, but I have no idea for the error records in the ProcessingStateMachine.

P.S.S: Ok. I think I will question now why we await that the ErrorRecord is committed. Lets say we got an error, we write the ErrorRecord. We already applied the blacklisting of the specific instance, which means we will ignore all further commands anyway. If a leader change happens and the error record was not committed and get lost, it means we will process the command again. TBH currently I see no value in awaiting the commit here. If this is really the case then we might be able to completely remove the await of certain positions (as written above the SnapshotDirector could await the processed position).

Reviewer 3: I can't think of a reason why we need to wait for the error record to be committed. :thinking: You might be right that we don't need to do it.

Author: Do you have an opinion on that @saig0 ?

Author: Would appreciate your input here @npepinpe 

What we could do is to go with the solution one (written in the ZEP and tested via POC) and if we are done and have enough time evaluate whether it is possible to use the other approach?

Reviewer: In fact, all snapshot related things should be removed from the logstream, no?

Author: What do you mean with that? Be aware that the director is already in the Broker and we have a separate snapshot module. Do you have anything specific you mean?

Reviewer: No worries, it's a bit out of scope (unless it simplifies your life). Here I mean the state management stuff - it doesn't fit the responsibilities of the log stream module as far as I can tell. As far as the engine is concerned, it only operates on runtime, and doesn't care about the snapshot. The log stream is the abstraction it uses to read the Raft log, but also doesn't really need to care about snapshots. I think a better approach is to move the state management back into the broker, and we can then see if it makes sense to have a separate module for that as an abstraction over the snapshot store for the engine, a bit like how the log stream is an abstraction of the Raft log for the engine. Hope that helps. 

I don't think it's necessary here, but if it makes your life easier then that's something we could consider (maybe just the first step, moving the state management into the broker).

Author: Hm ok I thought we already did this, but I might be wrong :) :+1:

Reviewer: lol, we did! I totally missed that, when did that happen? :smile:

Author: I think when we refactored the snapshot replication. Lot of stuff was moved by @deepthidevaki I think

Reviewer: so there's no more snapshot stuff in the log stream? :heart_eyes:

Author: hehe correct :)

Reviewer: Now we just need to move the atomix log storage impl to the broker or a special purpose module :smile:

Author: I rewrote this section and replaced it with the mentioned approach https://github.com/zeebe-io/enhancements/pull/16/commits/d9d7e776661a6c167d4a2068cf7d497edad9ed60 

We already have some PR's which helps to go in this direction :)

## Stream Processor

In order to run the continuous replay on the follower, the StreamProcessor will support two modes. The first one is the normal mode, which means replay until the end and then start normal processing of commands. This mode is used on the Leader side. The second mode is replay continuously which is used on the follower side. The mode can be set on building the StreamProcessor. 

## Exporter

Exporters are running on Leader's only. In order to make sure that we get exporter updates on the followers, we need to distribute that information to them. 

We do that via a SBE message, which is sent by the Leader `ExporterDirector` regularly. The SBE message will look like the following: 

```xml
  <sbe:message id="1" name="ExporterPositions">
    <group id="0" name="positions">
      <field id="0" name="position" type="uint64" semanticType="long"/>

>       <data id="1" name="exporterId" type="varDataEncoding" semanticType="string"/>

Reviewer 2: Do you mean that we don't need to track the positions for every exporter?
Please explain in more detail why we could replace the exporter id with the partition id.

>       <data id="1" name="exporterId" type="varDataEncoding" semanticType="string"/>

Reviewer 3: Oops. Sorry. That was a mistake. I didn't think of multiple exporters. Ignore my comment. :face_with_head_bandage:

>       <data id="1" name="exporterId" type="varDataEncoding" semanticType="string"/>

Author: :+1:

          // but this makes it still simpler to implement we dont want
          // to do it from Follower-To-Leader switch
          new StreamProcessorPartitionStep(ReplayMode.CONTINUOUSLY), // runs the StreamProcessor in replay only mode
          new SnapshotDirectorPartitionStep(), // snapshot director which takes regulary snapshots
          new RocksDbMetricExporterPartitionStep(),
          new ExporterSatellitePartitionStep() // in order to receive exporter updates
          );
```

This should be seen as an example, how it is actually done in the end is implementation detail. The key point is that we need to install some components once, and need to restore/reset some services on transitioning to a new role. We need to make sure that services which depend on each other have the latest reference. Like the SnapshotDirector, should take a snapshot of the actual database object etc. We need to make sure that all services are closed correctly. The ZeebePartition need to support bootstrap steps, which are installed only once. It has to take them done, if the partition is closed.

### Follower Install Request Handling

> In order to keep it simple we handle new `InstallRequests` as a new Follower transition, which will cause to recreate all dependent services/components. Why an `InstallRequest` can happen, you can read about it [here](#re-init-follower-partition). How the recreation look like, is again more of an implementation detail. It can be done by calling the `RoleTransitionListeners` again on the Raft level. This approach was used in our [POC](https://github.com/camunda-cloud/zeebe/issues/7328#issuecomment-867683162) and seems to be the simplest one.

Author: Thanks! I added it as unresolved question https://github.com/zeebe-io/enhancements/pull/16/commits/a495d741f21cc247a99be7be35ffdba66a1ec6ec

       
## Compatibility

We see no issues with this proposal.

### Rolling updates

We can have various starting scenarios, like followers have the same log, one follower is lagging behind, all followers are lagging behind. When thinking about rolling updates we need to consider how these different scenarios affect the procedure and outcome. 

#### Same state on all nodes

If we have the same state on all nodes and we start a rolling update we could start on a Leader or a Follower.

If we update first the Leader, a Follower will become next Leader, and the old Leader becomes Follower. The new Leader will replicate old entries, which should the Follower with the newest version be able to read in respect to backwards compatibility.

If we update first a Follower, it is similar to above.

#### Follower is lagging behind

> If one follower is lagging behind it is likely to happen that if the other Follower or Leader is updated that one of them becomes Leader again. This means we have an old Follower and maybe a Leader with a newer version. In our processing and replay state machine we have event filters, which verify the version of the read entry. Newer entries are ignored by the Follower with the old version. This means for a short period of time the Follower, with the old version, will not replay. This should be fine, since we expect that the rolling update is in a limited time frame and this follower wouldn't take over anyway (since it is lagging behind). 

Reviewer 2: > In our processing and replay state machine we have event filters, which verify the version of the read entry. 

Yes, we have a filter in the processing and the replay state machine. But it's good to know that it checks the version of the protocol (defined in `/protocol/pom.xml`) and not the version of the broker itself. 

Currently, the protocol version is not updated strictly. For example, when adding a new value type to the enum.

Apart from changing the protocol definition itself, we might need to consider other changes in the broker. For example, adding a new BPMN element type. If an old follower is not aware of a new BPMN element then it is not able to produce the correct state.

We should think about checking the broker version instead.

> Newer entries are ignored by the Follower with the old version.

Ignored is maybe not the correct word :sweat_smile: The version check throws an exception. I'm not sure how nice this looks in the logs. 

Maybe, we want to handle this case more clear to identify it in the logs when it happens.

Author: Thanks I created https://github.com/camunda-cloud/zeebe/issues/7449 for that

  * In general, we should verify that fail-over works as expected and that the followers are building there state correctly, such that a new Leader can use it.
     * Similar to above.
  * After fail-over the new Leader should be able to export after the last lowest exporter position.
  * After fail-over the new Leader should be able to process after the last processed command.
  * InstallRequest's should be properly handled by followers, and they should continue with applying events after restoring the state.
    * Maybe we verify whether new snapshots are taken, which is a good indication that it still makes progress.

### E2E

 In the E2E test we should verify similar things:

 * We should verify that fail-over work as expected and that the followers are building there state correctly. The leaders should be able to continue where the followers left of and there should be no issue on exporting records from the new leader.

### Benchmarks

Verify the improvements in process execution latency during fail-over and also the latency from Follower-To-Leader processing transition, which was the main goal.

>  * New metric we should add here is the time between receiving the LEADER Role event and starting to process, because this is really the time we can decrease with the ZEP. 

Author: I don't see that as requirement but yes it would be interesting. I added it https://github.com/zeebe-io/enhancements/pull/16/commits/e8286c4ea83bb1992665444f04718a6be7456543

### On becoming Leader

On fail-over the new Leader need to drain the map, before he can switch to normal mode. Furthermore, the dispatcher needs to be reset. This allows to continue with the current state, without reinstalling any dependencies.

### On becoming Follower

If, the node was previously Leader, then the writer strategy will be replaced to no longer really write the entries. Possible race conditions can be handled on the RAFT level, since it detects whether a follower tries to write an entry. The new follower continuous processing, with the next command (it skips events) and fills with the reader the corresponding map.

## Exporter Records

> Instead of sending the exported positions periodically over the wire, we could write the current state to the log. This can then be applied on the follower side to rebuild the exporter state, see related [issue](https://github.com/camunda-cloud/zeebe/issues/7088). This solution might be much more complex, since we need to handle here replay as well and this needs to be done in the `ExporterDirector` to keep the abstraction between the engine and exporters, which are part of the Broker.

Reviewer: :question: I don't understand how this is more complex than sending things over the network. Anything you send over the network may get corrupted, may arrive out of order, and you need to set up on both sides the sender/receiver. 

I get that you need to replay the log when starting the exporter, but it's much simpler replay than the engine, no? You just blindly apply the events from the last known position. But I admit you probably thought about this more than me, so I'm sure I'm missing something.

Author: Maybe complex is the wrong word. It is more effort to do it. It was really simple with the sbe message and sending it via the partition messaging service. Again here we can of course iterate over, but I currently see no need for it.

More details:

If you do it with the export records, then you need:

 * a new record type, value type intent etc.
 * a new record class
 * the exporter needs to write it, regularly (how timer or counts or both?)
 * the exporter is not allowed to export it
 * the exporter needs replay logic to restore the state, which means it starts from snapshot and goes through the complete log to search for the last exported positions, before it starts with normal exporting
 * the processor and replay need to skip export records as well
 * it makes the log longer; depending on the rate how often the records are written it might influence processing and exporting
 * the followers need to run also an exporter replay machine continuously
